src/main/kotlin/com/team4099/robot2023/subsystems/elevator/ElevatorIOSim.kt

```diff
@@ -0,0 +1,123 @@
+import org.team4099.lib.units.derived.IntegralGain
+import org.team4099.lib.units.derived.ProportionalGain
+import org.team4099.lib.units.derived.Volt
+import org.team4099.lib.units.derived.inVolts
+import org.team4099.lib.units.derived.volts
+import org.team4099.lib.units.perSecond
+
+object ElevatorIOSim : ElevatorIO {
+  val elevatorSim: ElevatorSim =
+    ElevatorSim(
+      DCMotor.getNEO(2),
+      ElevatorConstants.GEAR_RATIO,
+      ElevatorConstants.CARRIAGE_MASS,
+      ElevatorConstants.SPOOL_RADIUS,
+      ElevatorConstants.ELEVATOR_MAX_RETRACTION,
+      ElevatorConstants.ELEVATOR_MAX_EXTENSION,
+      ElevatorConstants.ELEVATOR_ANGLE,
+      true,
+    )
+
+  private val elevatorController =
+    PIDController(ElevatorConstants.SIM_KP, ElevatorConstants.SIM_KI, ElevatorConstants.SIM_KD)
+
+  init {
+
+    // Create a Mechanism2d display of an Arm with a fixed ArmTower and moving Arm.
+
+    val m_mech2d = Mechanism2d(90.0, 90.0)
+    val midNodeHome = m_mech2d.getRoot("Mid Node", 27.83, 0.0)
+    val MidNode =
+      midNodeHome.append(
+        MechanismLigament2d("Mid Cone Node", 34.0, 90.0, 10.0, Color8Bit(Color.kWhite))
+      )
+    val highNodeHome = m_mech2d.getRoot("High Node", 10.58, 0.0)
+    val HighNode =
+      highNodeHome.append(
+        MechanismLigament2d("High Cone Node", 46.0, 90.0, 10.0, Color8Bit(Color.kWhite))
+      )
+    val gridHome = m_mech2d.getRoot("Grid Home", 49.75, 0.0)
+    val GridNode =
+      gridHome.append(
+        MechanismLigament2d("Grid Wall", 49.75, 180.0, 50.0, Color8Bit(Color.kWhite))
+      )
+    val dsHome = m_mech2d.getRoot("Double Substation Home", 49.75, 37.0)
+    val DSRamp =
+      dsHome.append(
+        MechanismLigament2d(
+          "Double Substation Ramp", 13.75, 180.0, 10.0, Color8Bit(Color.kWhite)
+        )
+      )
+
+    val m_bumper =
+      gridHome.append(MechanismLigament2d("Bumper", 30.5, 0.0, 60.0, Color8Bit(Color.kRed)))
+
+    SmartDashboard.putData("Arm Sim", m_mech2d)
+  }
+
+  override fun updateInputs(inputs: ElevatorIO.ElevatorInputs) {
+    elevatorSim.update(Constants.Universal.LOOP_PERIOD_TIME.inSeconds)
+
+    inputs.elevatorPosition = elevatorSim.positionMeters.meters
+    inputs.elevatorVelocity = elevatorSim.velocityMetersPerSecond.meters.perSecond
+
+    inputs.leaderTempCelcius = 0.0.celsius
+    inputs.leaderStatorCurrent = 0.0.amps
+    inputs.leaderSupplyCurrent = elevatorSim.currentDrawAmps.amps / 2
+    inputs.leaderAppliedOutput = 0.0
+
+    inputs.followerTempCelcius = 0.0.celsius
+    inputs.followerStatorCurrent = 0.0.amps
+    inputs.followerSupplyCurrent = elevatorSim.currentDrawAmps.amps / 2
+    inputs.followerAppliedOutput = 0.0
+
+    RoboRioSim.setVInVoltage(
+      BatterySim.calculateDefaultBatteryLoadedVoltage(elevatorSim.currentDrawAmps)
+    )
+  }
+
+  override fun setOpenLoop(percentOutput: Double) {
+    elevatorSim.setInputVoltage(RoboRioSim.getVInVoltage() * percentOutput)
```

Comment:
should switch to using volts? im anti percentOutput because although it's right, this line isn't very intuitive tbh